code/gateway/gatewaylib.q

```diff
@@ -0,0 +1,105 @@
+//functionality loaded in by gateway
+//functions include: getserverids, getserveridstype, getserverscross, buildcross
+
+\d .gw
+
+getserverids:{[att]
+  if[99h<>type att;
+   // its a list of servertypes e.g. `rdb`hdb
+   // check if user attributes are a symbol list
+   if[not 11h=abs type att;
+     '" Servertype should be given as either a dictionary(type 99h) or a symbol list (11h)"
+   ];
+   servertype:distinct att,();
+   //list of active servers
+   activeservers:exec distinct servertype from .gw.servers where active;
+   //list of all servers
+   allservers:exec distinct servertype from .gw.servers;
+   activeserversmsg:". Available servers include: ",", " sv string activeservers;
+   //check if a null argument is passed
+   if[any null att;'"A null server cannot be passed as an argument",activeserversmsg];
+   //if any requested servers are missing then:
+   //if requested server does not exist, return error with list of available servers
+   //if requested server exists but is currently inactive, return error with list of available servers
+   if[count servertype except activeservers;
+    '"the following ",$[max not servertype in allservers;
+     "are not valid servers: ",", " sv string servertype except allservers;
+     "requested servers are currently inactive: ",", " sv string servertype except activeservers
+    ],activeserversmsg;
+   ];
+   :(exec serverid by servertype from .gw.servers where active)[servertype];
+  ];
+
+  // its a dictionary of attributes
+  // check if attribute types are correct types
+  correctAttTypes:`date`servertype`tables!14 11 11h;
```

Comment:
The rest of this file is all lower case for function names - can you stick to that